oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -0,0 +1,169 @@
+            }
+            return t;
+        });
+    }
+
+    private CompletableFuture<AccessTokenCapsule> refreshAccessToken(AccessTokenCapsule token) {
+        if (token.refreshToken() != null) {
+            // try refreshing token if refresh token was previously provided
+            try {
+                return refreshRequest.make(token);
+            } catch (TokenRequestException e) {
+                // token refresh request failed
+                // try to re-obtain access token
+                return obtainAccessToken(token);
+            }
+        }
+        // try to re-obtain access token
+        return obtainAccessToken(token);
+    }
+
+    private CompletableFuture<AccessTokenCapsule> getOrRefreshAccessToken(AccessTokenCapsule token,
+                                                                          boolean reset,
+                                                                          boolean lock) {
+        // check if it's still valid
+        final Instant instant = Instant.now().plus(refreshBefore);
+        if (token.isValid(instant)) {
+            // simply return a valid token
+            if (reset) {
+                return CompletableFuture.completedFuture(token).thenApply(t -> {
+                    tokenRef.set(t); // reset the token reference
+                    return t;
+                });
+            } else {
+                return CompletableFuture.completedFuture(token);
+            }
+        } else {
+            if (lock) {
+                synchronized (tokenRef) {
+                    // refresh token exclusively
+                    return refreshAccessTokenExclusively(instant);
+                }
+            } else {
+                // refresh token exclusively
+                return refreshAccessTokenExclusively(instant);
+            }
+        }
+    }
+
+    private CompletableFuture<AccessTokenCapsule> refreshAccessTokenExclusively(Instant instant) {
+        // after acquiring the lock, re-check if it's a valid token
+        final AccessTokenCapsule token = tokenRef.get();
+        if (token.isValid(instant)) {
+            // simply return a valid token
+            return CompletableFuture.completedFuture(token);
+        }
+        // otherwise, refresh it
+        return refreshAccessToken(token).thenApply(t -> {
+            tokenRef.set(t); // reset the token reference
+            if (tokenConsumer != null) {
+                tokenConsumer.accept(t); // store token to an optional storage (e.g. secret store)
+            }
+            return t;
+        });
+    }
+
+    @Override
+    public CompletionStage<AccessTokenCapsule> getAccessToken() {
+        AccessTokenCapsule token = tokenRef.get();
+        if (token != null) {
+            // token already present
+            return getOrRefreshAccessToken(token, false, true);
+        }
+        // token is not yet present
+        // lock and obtain token exclusively
+        synchronized (tokenRef) {
```

Comment:
`obtainAccessToken()` is an asynchronous operation, so I guess just using `synchronized` will not prevent a subclass from obtaining tokens simultaneously. Am I missing something? Also, it'd be nice if we can reduce the potential lock contention which may block the event loop.

Reply:
You are right! This was wrong. Changed the implementation to use a single-thread executor to obtain new or refresh existing token.